### PR TITLE
issue 21 -- Configuration Registry initial implementation

### PR DESCRIPTION
**Pull request by NorthWindH**
Initial implementation of configuration registry ( aside from load/store ) is ready for peer review. I tried to follow style guidelines exhibited in other files as much as possible.

files of note:
`src/gladius-src.js`
`src/core/Configurator.js`
`test/core/Configurator.js`

Please note the huge hack in src/core/Configurator.js:148 and the associated hack in test/core/Configurator.js:118, ideas on a better way to test construction/defaultConfig loading?

Issue thread: https://github.com/alankligman/gladius/issues/21

### CODE REVIEW

File: src/core/Configurator.js [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    
+    /* Configuration Node
+     * 
+     * Structure for storing configuration settings
+     */
+    var ConfNode = function( name, parent ) {
+        
+        var self = this;
+        
+        self.name = name;
```
Comment (humphd) [inline src/core/Configurator.js]: Use this vs. self for attaching instance properties.  Stashing this on self is a good idea, but don't use it everywhere.

File: src/core/Configurator.js [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+                }
+            }
+        });
+        
+        // Traverse the node tree given a path
+        self.traverse = function( path, doCreatePath ) {
+            
+            var rv;
+            
+            if ( path.length == 1 && path.charAt( 0 ) == '/' ) {
```
Comment (humphd) [inline src/core/Configurator.js]: === vs. ==

File: src/core/Configurator.js [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+                rv = self;
+            } else {
+                
+                // Parse path and traverse the node tree
+                var pathElems = path.split('/');
+                var curNode = self;
+                var successful = true;
```
Comment (humphd) [inline src/core/Configurator.js]: var a,
  b,
  c;

vs. using var var var

File: src/core/Configurator.js [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+                for ( var i = 0; successful && i < pathElems.length; ++ i ) {
```
Comment (humphd) [inline src/core/Configurator.js]: cache pathElems.length:

for (var i=0, iMax = pathElems.length; i < iMax; ++i)

File: src/core/Configurator.js [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+            return rv;
+        };
+        
+        // Notifies us that a value stored somewhere in the subtree rooted by
+        // this node has changed.
+        self.notify = function ( childName, path, newVal ) {
+            
+            if ( self.parent ) {
+                // Clean last slash if present
+                if ( path.length == 1 && path.charAt( 0 ) == '/' ) {
```
Comment (humphd) [inline src/core/Configurator.js]: === vs. ==

File: src/core/Configurator.js [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+                    path = '';
+                }
+                
+                // Build up the path
+                path = '/' + childName + path;
+                
+                // Call all of our listeners
+                for ( var key in self.listeners ) {
```
Comment (humphd) [inline src/core/Configurator.js]: when possible, avoid for..in loops, as they are more expensive.

File: src/core/Configurator.js [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+                    // is hasOwnProperty desired here? Seems unnecessary
```
Comment (humphd) [inline src/core/Configurator.js]: I would, since you don't want to work with keys in the prototype.

File: src/core/Configurator.js [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+        };
+    };
+    
+    /**
+     * More or less unique id generator
+     *
+     * Borrowed from Sagi Shkedy's Technical Blog
+     * http://blog.shkedy.com/2007/01/createing-guids-with-client-side.html
+     */
+    var _incCounter = 0;
```
Comment (humphd) [inline src/core/Configurator.js]: Should code for getting a guid not live somewhere more central in gladius?

File: src/core/Configurator.js [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+     * changes in configuration subtrees.
+     *
+     * Initially written by Hasan (northWind) Kamal-Al-Deen
+     */
+    var Configurator = function( defaultConfiguration ) {
+        
+        var self = this;
+        
+        var init = function() {
+            var self = this;
```
Comment (humphd) [inline src/core/Configurator.js]: not necessary, since self exists in the closure around init.

File: src/core/Configurator.js [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+            self.set = function( path, value ) {
+                var targetNode = self.node.traverse( path, true );
+                
+                targetNode.value = value;
+            };
+            
+            // Update configuration with given json object
+            self.update = function( json ) {
+                for ( var key in json ) {
+                    if (json.hasOwnProperty( key )) {   // Performance Note: perhaps protecting against the prototype is not required?
```
Comment (humphd) [inline src/core/Configurator.js]: leave this if doing for...in

File: src/core/Configurator.js [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+                        self.set( key, json[key] );
+                    }
+                }
+            };
+            
+            // Get a new configurator client for a node reachable using the given path.
+            // If provided, associate listenerFunc with the newly created configurator client.
+            self.getPath = function( path, listenerFunc ) {
+                var targetNode = self.node.traverse( path, true ),
+                    rv = new init();
```
Comment (humphd) [inline src/core/Configurator.js]: init is a ctor?  That's a bad name then.  ctor functions shoudl start with a capital.

File: src/core/Configurator.js [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+            self.listen = function( listenerFunc ) {
+                if ( listenerFunc ) {
+                    self.ignore();
+                    
+                    self.node.listeners[self.id] = listenerFunc;
+                }
+            };
+        };
+        
+        init.apply(self);
```
Comment (humphd) [inline src/core/Configurator.js]: apply is slower than using a direct function call, can you get around this?

File: src/core/Configurator.js [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+        
+        self.node = new ConfNode( 'ROOT', undefined );  // Associate client front end with tree root
```
Comment (humphd) [inline src/core/Configurator.js]: don't pass undefined, just leave the arg out: new ConfNode( 'ROOT' );

File: src/core/Configurator.js (hunk 6 above)
Comment (NorthWindH) [inline src/core/Configurator.js]: In this case, how do we perform the same operation? `self.listeners` is an associative array keyed with string ids. Should I be using `Object.keys()`? Is that not just as expensive? Is there a quicker alternative?

File: src/core/Configurator.js (hunk 8 above)
Comment (NorthWindH) [inline src/core/Configurator.js]: I'm inclined to think that it should. From what I could tell ( I will preemptively declare that I did not put much time into looking :) ), such code does not already exist in an accessible location.

File: src/core/Configurator.js (hunk 9 above)
Comment (NorthWindH) [inline src/core/Configurator.js]: Are you sure? In [src/core/Configurator.js:184](https://github.com/northWind87/gladius/blob/develop_issue_21/src/core/Configurator.js#L184), I use `new init()` to use init as a constructor of sorts. In this case, I want `self` to refer to the new object, not the object onwhich `getPath()` is running.

edit: `init()` removed

**Comment by NorthWindH** [comment]
Resolved all comments.

[Diff of changes](https://github.com/northWind87/gladius/compare/02688016...82292e15909e792812a312032d1788cfd8d9e8a6)

File: src/core/Configurator.js (hunk 8 above)
Comment (modeswitch) [inline src/core/Configurator.js]: There's already code for doing this. Use engine.nextGUID. You're not serializing it, so it should be safe.

File: src/core/Configurator.js [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+     */
+    var Configurator = function( defaultConfiguration, node ) {
+        
+        defaultConfiguration = defaultConfiguration || {};
+        
+        var self = this;
+        
+        this.node = node || new ConfNode( 'ROOT' );
+        
+        this.constructor = Configurator;    // TODO: FIXME Dirty Dirty Hack
```
Comment (modeswitch) [inline src/core/Configurator.js]: Why is this necessary?

File: src/core/Configurator.js (hunk 8 above)
Comment (NorthWindH) [inline src/core/Configurator.js]: sounds good, I will update the configurator to be partially constructing in order to allow this.

edit: [done](https://github.com/northWind87/gladius/blob/develop_issue_21/src/core/Configurator.js#L145)

File: src/core/Configurator.js (hunk 14 above)
Comment (NorthWindH) [inline src/core/Configurator.js]: @alankligman
This is actually a problem that I ran into during testing. Essentially, in order to test loading a JSON object during construction, I needed a way to create a configurator object on demand inside the test ([L118](https://github.com/alankligman/gladius/pull/52/files#L2R118)). At the moment, the constructor is not stored in a good place. Would it be a good idea to store a reference to the configurator constructor somewhere on engine? Perhaps the test should be changed? What do you think.

Comment (modeswitch) [inline src/core/Configurator.js]: You can use `Object.getPrototypeOf( engine.configurator ).constructor` to get Configurator, which you can then use to make a new instance by `var myConfigurator = new Configurator( engine )`.

Comment (NorthWindH) [inline src/core/Configurator.js]: Nice, that sounds perfect. I'll complain if I blow something up.

edit: Resolved. Updated construction method, now partially constructed.

**Comment by NorthWindH** [comment]
Ready for more review. Also looks like I've angered the end of line gods... my apologies.

edit: alright cleaned up, new diff

[diff of changes](https://github.com/northWind87/gladius/compare/82292e15909e792812a312032d1788cfd8d9e8a6...aa755ccd1215f76c13518cf6d651a3b14a542515)

**Comment by NorthWindH** [comment]
Implemented load, store, getJSON, clear
At the moment, load only loads from local storage ie a cookie

Review desired. I'll also be posting up some questions pointing to the code shortly.

File: src/core/Configurator.js [hunk 15]
```diff
@@ -0,0 +1,412 @@
...
+            }
+
+            var loadedJSON;
+            if ( URL ) {
+                // unimplemented
+            } else {
+                // TODO HACK this is extremely dangerous
+                // Do we have any dependencies that handle JSON parsing?
+                // Do browsers reliably provide native JSON parsing?
+                loadedJSON = eval(
```
Comment (NorthWindH) [inline src/core/Configurator.js]: In the words of check-lint, eval is evil. Do we have better options for parsing JSON?

File: src/core/Configurator.js [hunk 16]
```diff
@@ -0,0 +1,412 @@
...
+     *                 newConfInstance.update( myJSON );
+     *
+     * Initially written by Hasan (northWind) Kamal-Al-Deen
+     */
+    var Configurator = function( engine, defaultConfiguration ) {
+        
+        defaultConfiguration = defaultConfiguration || {};
+
+        var self = this,
+            _gladiusCookieName = 'gladius_registry',
```
Comment (NorthWindH) [inline src/core/Configurator.js]: Cookie name for local storage is fixed in stone. Is this OK?

File: src/core/Configurator.js [hunk 17]
```diff
@@ -0,0 +1,412 @@
...
+            _gladiusCookieLifetime = 365,
+
+            // TODO: These should be moved into the networking system
+            // Cookie manipulation functions, slightly modified.
+            // Originally found on Peter-Paul Koch's site, quirksmode
+            // http://www.quirksmode.org/js/cookies.html
+            createCookie = function(name,value,days) {
```
Comment (NorthWindH) [inline src/core/Configurator.js]: Cookie manipulation functions borrowed from inter-webs. Is having them here OK? Should they be abstracted to another module?

File: src/core/Configurator.js (hunk 15 above)
Comment (JSilver99) [inline src/core/Configurator.js]: Isn't there JSON.parse(str) for that? eval parses any JavaScript, and you probably don't want that.
https://developer.mozilla.org/en/JSON

Comment (NorthWindH) [inline src/core/Configurator.js]: @JSilver99
Yeah I'm quite wary of using eval; this is code that can't go into the mainline. With respect to JSON.parse(), how cross-browser compatible is it? I could not find a source that definitively states where JSON is available.

Comment (JSilver99) [inline src/core/Configurator.js]: It looks like any recent stable release of any major browser accepts it.
A quick google brought up this page: http://caniuse.com/json

Comment (NorthWindH) [inline src/core/Configurator.js]: Done and done! da80e6a8d47984d92fba6822ec78e5c712f55c62

File: src/core/Configurator.js [hunk 18]
```diff
@@ -0,0 +1,416 @@
...
+
+define( function ( require ) {
+    
+    /* Configuration Node
+     * 
+     * Structure for storing configuration settings
+     */
+    var ConfNode = function( name, parent ) {
+        
+        var self = this;
```
Comment (modeswitch) [inline src/core/Configurator.js]: Use 'that' to store the current context instead of 'self'.

File: src/core/Configurator.js [hunk 19]
```diff
@@ -0,0 +1,416 @@
...
+                    rv = curNode;
+                }
+            }
+            
+            return rv;
+        };
+
+        // Serializes this node and all of its children as JSON
+        this.getJSON = function() {
+            var rv = {}, children = this.children;
```
Comment (modeswitch) [inline src/core/Configurator.js]: Newline after comma.

File: src/core/Configurator.js [hunk 20]
```diff
@@ -0,0 +1,416 @@
...
+                }
+
+                this.parent.notify( path, newVal, oldVal );
+            }
+        };
+        
+        // Traverse the node tree given a path
+        this.traverse = function( path, doCreatePath ) {
+            
+            var rv;
```
Comment (modeswitch) [inline src/core/Configurator.js]: Give this variable a better name (more descriptive).

File: src/core/Configurator.js [hunk 21]
```diff
@@ -0,0 +1,416 @@
...
+            return rv;
+        };
+    };
+    
+    /* Configurator
+     *
+     * Loads and stores configuration data. Allows external code to listen for
+     * changes in configuration subtrees.
+     *
+     * In Gladius, a configurator can be obtained in 3 ways:
```
Comment (modeswitch) [inline src/core/Configurator.js]: Remove the rest of this comment below this line.

File: src/core/Configurator.js (hunk 16 above)
Comment (modeswitch) [inline src/core/Configurator.js]: The configurator should accept the cookie name as a parameter during construction.

File: src/core/Configurator.js (hunk 17 above)
Comment (modeswitch) [inline src/core/Configurator.js]: Refactor this into a separate module. You can put the code in src/common.

File: test/core/Configurator.js [hunk 22]
```diff
@@ -0,0 +1,502 @@
...
+  nomen: false */
+/*global gladius: false, document: false, window: false, module: false, start,
+  test: false, expect: false, ok: false, notEqual: false, stop, QUnit: false */
+
+/***
+ * Configurator Test Suite
+ *
+ * A test suite for the Configurator subsystem.
+ *
+ * Initially written by Hasan (northWind) Kamal-Al-Deen
```
Comment (modeswitch) [inline test/core/Configurator.js]: Remove this.

File: src/core/Configurator.js (hunk 16 above)
Comment (NorthWindH) [inline src/core/Configurator.js]: On construction, configurator currently accepts a single defaultConfiguration value which is interpreted as a registry JSON object and loaded into the configurator immediately. How about we make the configurator accept a standard options-style object with defaultConfiguration, cookieName and cookieLifetime properties?

Comment (modeswitch) [inline src/core/Configurator.js]: That makes sense. Bake a minimal default configuration into the engine that we can pass in to the configurator.

Comment (NorthWindH) [inline src/core/Configurator.js]: sounds good.

**Comment by NorthWindH** [comment]
Implemented all suggestions up until now.

File: src/common/Cookie.js [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+        }
+        return null;
+    };
+
+    this.eraseCookie = function(name) {
+        createCookie(name,"",-1);
+    };
+};
+
+window.gladiusCookie = new _Cookie();
```
Comment (NorthWindH) [inline src/common/Cookie.js]: window.cookie ( window.math style ) is perhaps vague, too similar to document.cookie
window.gladiusCookie is perhaps not the best name in the world but its not terrible, is a better name desired?

File: test/core/Configurator.js [hunk 24]
```diff
@@ -0,0 +1,500 @@
...
+        var jsonActual = config.getJSON();
+        for ( var i = 0, maxlen = expectedKeys.length; i < maxlen; ++i ) {
+            equal(
+                jsonActual[expectedKeys[i]],
+                jsonExpected[expectedKeys[i]]
+            );
+        }
+    });
+
+    // Test clear
```
Comment (NorthWindH) [inline test/core/Configurator.js]: All of the tests from here on-ward have no comment as their role is relatively obvious; I don't like this but I don't hate it enough to write a comment for each of 50+ tests. This can be changed if desired. Please indicate preference.

File: src/gladius-src.js [hunk 25]
```diff
@@ -36,7 +37,11 @@ define( function ( require ) {
         subsystems, prop;
 
         this.options = options || {};
+        if ( !this.options.defaultConfiguration )   this.options.defaultConfiguration = {};
+        if ( !this.options.cookieName )             this.options.cookieName = 'gladius_registry';
```
Comment (NorthWindH) [inline src/gladius-src.js]: Threw cookieName and cookieLifetime onto gladius.options. Is this desirable?

File: src/common/Cookie.js (hunk 23 above)
Comment (modeswitch) [inline src/common/Cookie.js]: Why is this being stored on the window instance?

File: test/core/Configurator.js (hunk 24 above)
Comment (modeswitch) [inline test/core/Configurator.js]: Each of the tests (equal, ok, etc) needs a string associated with it that makes clear the intent of the test. That string is the only information available on the test page if the tests fail.

File: src/common/Cookie.js (hunk 23 above)
Comment (NorthWindH) [inline src/common/Cookie.js]: Followed the style of Math.js, where is the preferred storage location.

File: test/core/Configurator.js (hunk 24 above)
Comment (NorthWindH) [inline test/core/Configurator.js]: No problem.

File: src/common/Cookie.js (hunk 23 above)
Comment (modeswitch) [inline src/common/Cookie.js]: That's not the correct thing to do here. If I create multiple engines in the same page, this code will break.

File: src/core/Configurator.js [hunk 26]
```diff
@@ -0,0 +1,374 @@
...
+            for ( var childKey in children ) {
+                if ( children.hasOwnProperty( childKey ) ) {
+                    children[childKey].clear();
+                }
+            }
+        };
+
+        // Builds a parent path for this node
+        this.getParentPath = function() {
+            rv = '';
```
Comment (modeswitch) [inline src/core/Configurator.js]: Please fix these variable names (and similar ones) so that they are clearer.

File: src/gladius-src.js (hunk 25 above)
Comment (modeswitch) [inline src/gladius-src.js]: The cookie name can go into the default config. Also, don't assign a default cookie name. If one isn't provided, then we will avoid storing the registry.

**Comment by NorthWindH** [comment]
Whew! All suggestions thus far brought up have been implemented and as a result, review requested!

## Details
- All suggested changes upto now have been implemented, see commit messages for details.
- Configurator now uses indexeddb as load/store backend.
- Code has been brought up to date, ack/develop merged into incoming branch.

File: src/gladius-src.js [hunk 27]
```diff
@@ -130,10 +143,12 @@ define( function ( require ) {
                 this.options.setup( this );
             }
 
-            // Let caller know the engine instance is ready.
-            if (callback) {
-                callback(this);
-            }
+            // Load any persistent configuration then let caller know the engine instance is ready.
```
Comment (modeswitch) [inline src/gladius-src.js]: The config needs to be loaded earlier than this. All services are started by this point. Should be the first thing to do, even before _math is created.

**Comment by modeswitch** [comment]
There's a merge conflict in gladius-src, you will need to sync from develop.

**Comment by NorthWindH** [comment]
Review requested.
Merge conflict resolved.
Configurator now loads before anything else.

**Comment by NorthWindH** [comment]
Configurator dies if db is not allocated. This behaviour should be amended to not use a db instead.

**Comment by NorthWindH** [comment]
Engine no longer provides nextGUID, Configurator should require lang.js and use window.guid

**Comment by NorthWindH** [comment]
rootConfNode shouldn't be stored on engine, pass the root configurator in as a constructor parameter instead

**Comment by NorthWindH** [comment]
Have load accept named parameter list

**Comment by NorthWindH** [comment]
Review requested.
Identified problems resolved.

File: test/core/configurator.js [hunk 28]
```diff
@@ -0,0 +1,602 @@
...
+/***
+ * Configurator Test Suite
+ *
+ * A test suite for the Configurator subsystem.
+ */
+
+(function() {
+
+    var engine = null,
+        getNoDBMsg = function() {
```
Comment (modeswitch) [inline test/core/configurator.js]: Why is this a function? Also, there's already a mechanism to start up a server to run the tests (make test) so just instruct people to use that.

File: test/core/configurator.js [hunk 29]
```diff
@@ -0,0 +1,602 @@
...
+
+        msg = function( key, val ) { return 'Cleared /foo branch, value for key ' + key + ' should be ' + ( val ? val : 'empty' )};
+        equal( config.get( key1 ), '', msg( key1, '' ) );
+        equal( config.get( key2 ), '', msg( key2, '' ) );
+        equal( config.get( key3 ), val3, msg( key3, val3 ) );
+    });
+
+    // Test load/store to/from local storage
+    test( 'local store/load', function() {
+        // Do not perform these tests if db access is not available
```
Comment (modeswitch) [inline test/core/configurator.js:428]: Don't disable these tests. We want to see when they fail!

**Comment by modeswitch** [comment]
Make sure to test the case where DB access in unavailable to make sure your flag is set.

File: test/core/configurator.js (hunk 28 above)
Comment (NorthWindH) [inline test/core/configurator.js]: Also change function to a string.

**Comment by NorthWindH** [comment]
Changes implemented.
Review requested.

**Comment by NorthWindH** [comment]
load/store and merge store tests may be passing when there is no db connectivity, this is bad behaviour

I could not reproduce this on my machine, @AlanKligman plz confirm? thanks! --@northWind87

**Comment by modeswitch** [comment]
Tests are not running after merging your latest changes. Can you confirm that they run and pass for you?

**Comment by NorthWindH** [comment]
@alankligman Hmmm I did some tests, the results are interesting.

### FF7
- **northwind/develop_issue_21:** works!
- **ack/develop:** works!

### FF8
- **northwind/develop_issue_21:** Broken. Configurator attempts to load, fails to acquire db, then calls gladius-src/initialize. Eventually threading/proxy throws an error to console ( "Could not get domain!" ) from threading/proxy.js Line 29
- **ack/develop:** Broken. No error message, did not investigate.

### Nightly ( firefox 11.0a1 (2011-12-09) )
- **northwind/develop_issue_21:** Broken. Configurator fails to initialize db ( "setVersion is not a function" ). Needs try/catch statement.
- **ack/develop:** Works!

Which version should we be aiming to please? :S

**Comment by NorthWindH** [comment]
**northwind/develop_issue_21** now works in nightly! setVersion is now nolonger a function so the load/store tests fail appropriately while the rest of the suite runs with no errors.

**Comment by modeswitch** [comment]
If it works in FF9 that's good. It will need to work in nightly as well at some point, but that can be a different task.

**Comment by NorthWindH** [comment]
I've tested the configurator branch in FF9; the test suite is not broken but the configurator load/store tests obviously don't pass because setVersion is no longer a function. I'll look into how the interface has changed and modify appropriately.

Also, the core/thread: Basic test never completes on the configurator branch on FF9. I've tested the ack/develop branch on FF9 and it did not have this issue. I traced the test code around a little on the configurator branch but didn't find any obvious problems. The last thing that happens is that the function ready() on pool.js gets called and goes down the path **( _queuedRequests.length > 0 ) = false => _readyThreads.push( thread )**

**Comment by NorthWindH** [comment]
test suite should now run on FF9. Configurator tests should pass. @alankligman please confirm that tests run on your end?

You may have to clear the web storage for the tests page then reload the page. To do that, click the page's icon in the address bar, click more information, navigate to the permissions tab then click 'clear storage' under 'maintain offline storage'.

**Note:** core/thread Basic seems to pass after about... 5 minutes of waiting lol

File: src/core/configurator.js [hunk 30]
```diff
@@ -0,0 +1,525 @@
...
+            that = this,
+            canUseDB = true,
+
+            // Should dbVersion be picked up from default.js too?
+            dbVersion               = '1',
+            dbName                  = '',
+            objectStoreName         = '',
+            gameId                  = '',
+
+            _getErrorFunc = function( error, funcName, msg ) {
```
Comment (modeswitch) [inline src/core/configurator.js]: This function doesn't seem to do any useful work. Why not just concatenate the strings and pass them into the error handler?

File: src/core/configurator.js [hunk 31]
```diff
@@ -0,0 +1,525 @@
...
+                                // Delete object store if it's around
+                                if ( containsObjectStore ) {
+                                    db.deleteObjectStore( objectStoreName );
+                                }
+                                createObjectStore();
+                            };
+                        }
+                    }
+                } catch ( e ) {
+                    _getErrorFunc( error, '_ensureObjectStore', 'encountered exception' )( e );
```
Comment (modeswitch) [inline src/core/configurator.js]: This try block is way too large. Put it just around the code that might throw an exception and make the error message specific. "Encountered exception" is not acceptable.

File: src/core/configurator.js [hunk 32]
```diff
@@ -0,0 +1,525 @@
...
+            if ( listenerFunc ) {
+                targetNode.listeners[resultConf.id] = listenerFunc;
+            }
+
+            return resultConf;
+        };
+
+        // Remove listener currently associated with client.
+        this.ignore = function() {
+            var curListener = this.node.listeners[this.id];
```
Comment (modeswitch) [inline src/core/configurator.js]: This variable assignment is unnecesary.

File: src/core/configurator.js [hunk 33]
```diff
@@ -0,0 +1,525 @@
...
+        // Define read only access
+        var defReadOnly = function( name, val ) {
+            Object.defineProperty( that, name, {
+                get: function() {
+                    return val;
+                }
+            });
+        };
+
+        defReadOnly( 'KEY_GAME_ID', KEY_GAME_ID );
```
Comment (modeswitch) [inline src/core/configurator.js]: This is also unnecessary. Just define the properties.

File: src/core/configurator.js [hunk 34]
```diff
@@ -0,0 +1,525 @@
...
+            objectStoreName         = '',
+            gameId                  = '',
+
+            _getErrorFunc = function( error, funcName, msg ) {
+                return function( e ) {
+                    if ( error ) error( 'Gladius/Configurator-' + funcName + ': ' + msg + ( e ? ', error object: ' + e.toString() : '' ) );
+                };
+            },
+
+            _injectDB = function( options ) {
```
Comment (modeswitch) [inline src/core/configurator.js]: Rename this to openDB, or something a little more clear.

File: src/gladius-src.js [hunk 35]
```diff
@@ -25,7 +27,10 @@ define( function ( require ) {
     // exist, mainly gladius.ready from gladius.js. Check tools/wrap.start
     // for protections against overwriting an existing gladius in the page,
     // for when gladius is built for deployment.
-    global = window.gladius || ( window.gladius = {} );
+    global = window.gladius || ( window.gladius = {} ),
+
+    // Get testing game id ready
+    testGameID = window.guid();
```
Comment (modeswitch) [inline src/gladius-src.js]: This does not belong here. Remove it.

File: src/gladius-src.js [hunk 36]
```diff
@@ -34,120 +39,145 @@ define( function ( require ) {
...
+                if ( callback ) {
+                    callback( this );
+                }
+            }));
+        };
+
+        // Load persistent configuration then proceed with rest of engine startup
+        this.configurator.load( {
+            callback: initialize,
+            error: initialize
```
Comment (modeswitch) [inline src/gladius-src.js]: This needs to be fixed. Initialize won't handle errors that come from the configurator.

Comment (NorthWindH) [inline src/gladius-src.js]: At the moment, should load be unable to read from the db ie if db access is not available, the error handler is called. Since in this case the engine should load regardless, what would be the desired course of action?

**Comment by NorthWindH** [comment]
@alankligman 
Ready for review, everything fixed up.
